File: python-puka/receive_logs_direct.py
#!/usr/bin/env python
import puka
import sys

client = puka.Client("amqp://localhost/")
promise = client.connect()
client.wait(promise)


promise = client.exchange_declare(exchange='direct_logs', type='direct')
client.wait(promise)

promise = client.queue_declare(exclusive=True)
queue_name = client.wait(promise)['queue']

severities = sys.argv[1:]
if not severities:
    print >> sys.stderr, "Usage: %s [info] [warning] [error]" % (sys.argv[0],)
    sys.exit(1)

for severity in severities:
    promise = client.queue_bind(exchange='direct_logs', queue=queue_name,
                                routing_key=severity)
    client.wait(promise)


print ' [*] Waiting for logs. To exit press CTRL+C'

consume_promise = client.basic_consume(queue=queue_name, no_ack=True)
while True:
    msg_result = client.wait(consume_promise)
    print " [x] %r:%r" % (msg_result['routing_key'], msg_result['body'])

File: php-thesis/routing/receive_logs_direct.php
<?php

declare(strict_types=1);

use Thesis\Amqp\Client;
use Thesis\Amqp\Config;
use Thesis\Amqp\DeliveryMessage;
use function Amp\trapSignal;

require_once __DIR__.'/../vendor/autoload.php';

$client = new Client(
    new Config(
        urls: ['rabbitmq:5672'],
    ),
);

$channel = $client->channel();

$channel->exchangeDeclare(exchange: 'direct_logs', exchangeType: 'direct');
$queue = $channel->queueDeclare(exclusive: true);

$severities = array_slice($argv, 1);
if ($severities === []) {
    file_put_contents('php://stderr', "Usage: $argv[0] [info] [warning] [error]\n");
    exit(1);
}

foreach ($severities as $severity) {
    $channel->queueBind(queue: $queue->name, exchange: 'direct_logs', routingKey: $severity);
}

echo " [*] Waiting for logs. To exit press CTRL+C\n";

$channel->consume(
    callback: static function (DeliveryMessage $delivery): void {
        echo " [x] {$delivery->routingKey}: {$delivery->message->body} \n";
    },
    queue: $queue->name,
    noAck: true,
);

trapSignal([\SIGTERM, \SIGINT]);

$client->disconnect();
